README.md
# Library for EVA II PRO WiFi Smart Dehumidifier appliance
[![PyPI](https://img.shields.io/pypi/v/midea-inventor-lib.svg)](https://pypi.org/project/midea-inventor-lib/)
[![](https://img.shields.io/pypi/pyversions/midea-inventor-lib.svg)](https://pypi.org/project/midea-inventor-lib/)
[![](https://img.shields.io/pypi/l/midea-inventor-lib.svg)](https://pypi.org/project/midea-inventor-lib/)
[![](https://img.shields.io/pypi/wheel/midea-inventor-lib.svg)](https://pypi.org/pypi/midea-inventor-lib/)
[![](https://img.shields.io/pypi/status/midea-inventor-lib.svg)](https://pypi.org/pypi/midea-inventor-lib/)
[![](https://img.shields.io/pypi/implementation/midea-inventor-lib.svg)](https://pypi.org/pypi/midea-inventor-lib/)
[![<100kB](https://img.shields.io/github/languages/code-size/barban-dev/midea_inventor_dehumidifier.svg)](https://github.com/barban-dev/midea_inventor_dehumidifier)
[![<100kB](https://img.shields.io/github/repo-size/barban-dev/midea_inventor_dehumidifier.svg)](https://github.com/barban-dev/midea_inventor_dehumidifier)
[![Build status](https://travis-ci.com/barban-dev/midea_inventor_dehumidifier.svg?branch=master)](https://travis-ci.com/barban-dev/midea_inventor_dehumidifier)
[![Known Vulnerabilities](https://snyk.io/test/github/barban-dev/midea_inventor_dehumidifier/badge.svg?targetFile=requirements.txt)](https://snyk.io/test/github/barban-dev/midea_inventor_dehumidifier?targetFile=requirements.txt)
[![HitCount](http://hits.dwyl.io/barban-dev/midea_inventor_dehumidifier.svg)](http://hits.dwyl.io/barban-dev/midea_inventor_dehumidifier)
[![Donate](https://img.shields.io/badge/Donate-PayPal-green.svg)](https://www.paypal.com/cgi-bin/webscr?cmd=_s-xclick&hosted_button_id=5E7ULVFGCGKU2&source=url)


Author: Andrea Barbaresi =2018, 2019=

Licence: GPLv3

This repo contains the python package ***midea_inventor_lib*** that implements a client-side library to connect to the Web API provided by Midea/Inventor, in order to remotely control an **EVA II PRO WiFi Smart Dehumidifier device**.

Info about the dehumidifier appliance can be found [here.](https://www.inventorappliances.com/dehumidifiers/eva-ii-pro-wi-fi-20l)

You can buy the smart dehumidifier appliance (WiFi version) on Amazon (the two links below contain my referral code):
* [Amazon.co.uk](https://www.amazon.co.uk/gp/product/B07665CCSM/ref=as_li_qf_asin_il_tl?ie=UTF8&tag=barban0d-21&creative=6738&linkCode=as2&creativeASIN=B07665CCSM&linkId=a7408b12a09679586e1816acc3c9d74b)
* [Amazon.it](https://www.amazon.it/gp/product/B075486X31/ref=as_li_tl?ie=UTF8&camp=3414&creative=21718&creativeASIN=B075486X31&linkCode=as2&tag=barban03-21&linkId=33e8ff818aaa4b45f0c320e6661773b2)


Target devices
--------------
Even though the library has been designed to generically address any kind of existing MIDEA devices, ***please note that at the moment the implemented functionalities work on the dehumidifier appliance only (0xA1 type devices).***

If you are interested in developing code that is able to control Midea/Inventor Air Condition systems (0xAC type devices), you can have a look at ***midea-air-condition*** Ruby&Rails library by **Balazs Nadasdi** available [here.](https://github.com/yitsushi/midea-air-condition)


Prerequisites
-------------
In order to control the EVA II PRO WiFi Smart Dehumidifier appliance using the provided Python library, first of all it is necessary to download and install the official App, in order to register a valid user to the cloud platform (a valid email address is required). 
The official companion Apps are available on Google's and Apple's App Stores:
* [Google Play](https://play.google.com/store/apps/details?id=com.inventor)
* [Apple Store](https://itunes.apple.com/gr/app/invmate-ii/id1109243423)

Once connected with valid credentials (i.e. email address and password), your home device has to be added to the list of configured devices using the App (please refer to the manual of the official App to accomplish this task).

Once having a valid registered user and the home device configured, you can start to use the python library instead of the offical App to control the device via Internet (both the client when the library is installed and the home device should be connected to the Internet).


Installation
------------
Install from PyPi using [pip](http://www.pip-installer.org/en/latest), a package manager for
Python.
```
pip install midea-inventor-lib
```
Don't have pip installed? Try installing it, by running this from the
command line:
```
$ curl https://raw.github.com/pypa/pip/master/contrib/get-pip.py | python
```
Or, you can [download the source code (ZIP)](https://github.com/barban-dev/midea_inventor_dehumidifier/zipball/master) and then run:
```
python setup.py install
```
You may need to run the above commands with ``sudo``.


Getting started
---------------
Minimal steps to use the library in your python code are reported below:

**Step 1: Include the python package**
```python
from midea_inventor_lib import MideaClient
```
**Step 2: Instantiate the MideaClient object**

Using clear-text password:
```python
client = MideaClient("user.example@gmail.com", "myPassword", "")
```
Using password's sha-256 hash:
```python
client = MideaClient("user.example@gmail.com", "", "76549b827ec46e705fd03831813fa52172338f0dfcbd711ed44b81a96dac51c6")
```
**Enable logging (optional):**

You can enable logging by setting the 'verbose' parameter to True (default is False) in the MideaClient constructor. 
Set 'debug' parameter to True in order to log debugging messages too (default is False).
Set 'logfile' string parameter to a full-path filename in order to make the library log messages into a file instead of using the console (default).
E.g.:
```python
_email = "user@example.com"
_password = "passwordExample"
_sha256password = ""
_verbose = True		#Enable logging
_debug = False		#Disable debug messages
_logfile = ""		#Log to console (default)
client = MideaClient(_email, _password, _sha256password, _debug, _verbose, _logfile)
```
**Step 3: Activate a new session by logging in**
```python
res = client.login()
if res == -1:
  print "Login error: please check log messages."
else:
  sessionId = client.current["sessionId"]
```
**Step 4: Get the target deviceId by retrieving the list of configured appliances**
```python
appliances = {}
appliances = client.listAppliances()
for a in appliances:
  print "[id="+a["id"]+" type="+a["type"]+" name="+a["name"]+"]"
```
**Step 5: Send commands to control the target device**
Get the device state:
```python
res = client.get_device_status(deviceId)
if res == 1:
  print client.deviceStatus.toString()
```
Power-on:
```python
res = client.send_poweron_command(deviceId)
if res:
  print client.deviceStatus.toString();
```
Power-off:
```python
res = client.send_poweroff_command(deviceId)
if res:
  print client.deviceStatus.toString();
```
Set Ion on:
```python
res = client.send_ion_on_command(deviceId)
if res:
  print client.deviceStatus.toString();
```
Set Ion off:
```python
res = client.send_ion_off_command(deviceId)
if res:
  print client.deviceStatus.toString();
```
Set fan speed:
```python
if speed > 0 and speed < 100:
  res = client.send_fan_speed_command(deviceId, speed)
  if res:
    print client.deviceStatus.toString();
```
Set target humidity:
```python
if hum >= 30 and hum <= 70:
  res = client.send_target_humidity_command(deviceId, hum)
  if res:
    print client.deviceStatus.toString();
```
Set operation mode:
```python
if mode > 0 and mode < 5:
  res = client.send_mode_command(deviceId, mode)  #set Mode (1:TARGET_MODE, 2:CONTINOUS_MODE, 3:SMART_MODE, 4:DRYER_MODE)
  if res:
    print client.deviceStatus.toString();
```
Set updated status (usefull to update multiple attributes at one):
```python
status =client.get_device_status(deviceId)  #get current status
#Update status
status.ionSetSwitch = 1   #Ion on
status.setMode = 1        #target-mode
res = self._client.send_update_status_command(self._device["id"], status)
if res:
  print client.deviceStatus.toString();
```

Client example
--------------
This repo also contains a fully working client (***dehumi_control.py***) that demonstrates how to use the ***midea_inventor_lib*** library in order to control the EVA II PRO WiFi Smart Dehumidifier appliance via a Command Line Interface.

To use the client, the email address of a registered user and the associated password have to be provided via command line parameters (either clear-text password or password's sha-256 hash has to be provided using the '-p' or '-s' options):
```
# python dehumi_control.py  -h
Usage:dehumi_control.py -e <email_address> -p <cleartext password> -s <sha256_password> -l <logfile> [-h] [-v] [-d]
```

Home Assistant custom-component
-------------------------------
***[NEW]*** Custom-component for the Home Assistant platform can be found on ``/homeassistant`` folder.
In order to activate the component, follow these steps:

**Step 1: Copy necessary files to your HA's configuration shared folder** 

Copy ***the content of*** ``/homeassistant`` folder (***not the folder itself***), ***including subfolders***, on your HA's configuration shared folder.

Copy ``/midea_inventor_lib`` folder on ``\deps\lib\python3.6\site-packages\`` of your HA's configuration shared folder.

Final result should look similar to this:
```
\\<ha_ip_address>
    └── config
        ├── custom_components
        │   ├── climate
        │   │   └── midea_dehumi.py
        │   ├── midea_dehumi.py
        │   └── sensor
        │       └── midea_dehumi.py
        └── deps
            └── lib
                └── python3.6
                    └── site-packages
                        └── midea_inventor_lib
                            └── libfiles...
```

**Step 2: Activate midea_dehumi platform on your HA's configuration file**

Add the following section in your ``configuration.yaml``
```
midea_dehumi:
  username: user@example.com
  password: passwordExample
```
As usual, you can hide your secret password by means of ``!secret`` notation by specifing it in ``secrets.yaml``

Alternatively, if you prefer, ``sha256password`` parameter can be used instead of the ``password`` one to specify password's sha-256 hash
```
midea_dehumi:
  username: user@example.com
  sha256password: cf76d55503cdee3....
```
**Step 3: Activate DEBUG-level logging (optional)**

It is highly suggested to activate DEBUG-level logging for the three components the midea_dehumi platform consists of. This let you perform troubleshooting analysis if the component doesn't work as expected.
```
logger:
  default: info
  logs:
    custom_components.midea_dehumi: debug
    custom_components.climate.midea_dehumi: debug
    custom_components.sensor.midea_dehumi: debug
```
** Step 4: Restart HA**

You can restart HA by using one of your preferred method (e.g. using restart button on Settings>General>Server Management og HA's Web dashboard).

If everything is ok, you will find the following two new entities in your HA dashboard:

* **climate.midea_dehumi_*[Device_ID]***
* **sensor.midea_dehumi_*[Device_ID]*_humidity**

By means of the climate entity, you can control your appliance whereas the sensor reports the detected current humidity of your environment.

If you cannot find the entities reported above, check the logs generated by HA to track the issue.

Internals 
---------
***You can skip this part if you are not interested in technical details concerning the format of the API messages used by the library.***

Official companion Apps for Android and IOS platforms are based on the midea-SDKs made available by Midea Smart Technology Co., Ltd.:
* [ios-sdk](https://github.com/midea-sdk-org/ios-sdk)
* [android-sdk](https://github.com/midea-sdk-org/android-sdk)

According to the SDK's documentation, "MideaSDK is a software develop kit maintained by MSmart. You can develop your own APP, Smart Hardware or Smart TV based on this SDK to control the smart appliances produced by Midea."

Official documentation for the open API can be found here (chinese language only):
https://github.com/midea-sdk/midea-sdk.github.io/tree/master/api

Apart Androd and IOS platforms, no other environment is currently officially supported. In order to develop the client-side library for all the platform supporting Python, I used a Man-In-The-Middle Web Proxy as a packet sniffer to understand the basics on the API messages exchanged between the offical Android client and the Midea cloud Server.

Web API server can be reached via ```https://mapp-appsmb.com/<endpoint>``` (POST web requests shoud be used).

A brief description of the most relevant endpoints follows:

```/v1/user/login/id/get``` endpoint with 'loginAccount' parameter is used to get 'loginId' parameter (different for each session).

```/v1/user/login``` endpoint with 'password' parameter is used to perform the login ('accessToken' and 'sessionId' parameters are returned). The password parameter sent by the client is SHA-256 hash of a string derived from 'loginId', 'password' and 'appKey' parameters.

```/v1/appliance/user/list/get``` endpoint is used to retrieve the list of configured devices together withh all the associated parameters ('name', 'modelNumber', 'activeStaus', 'onlineStatus', etc.)


CHANGELOG.md
# midea_inventor_dehumidifier Change Log

All notable changes to this project will be documented in this file.

## [1.0.5] - 2019-01-06
### Fixed
- Fixed bug preventing Home Assistant's sensor entity to update current humidity value

## [1.0.4] - 2019-01-06
### Fixed
- Fixed faulty getter methods due to typos for MideaDehumidificationDevice class
### Added
- Implemented send_update_status_command method for MideaClient class
- Added custom_component to support new midea_dehumi platform in Home Assistant 

## [1.0.3] - 2018-12-30
### Added
- Added support for cached device status results
- Added support for invalidSession error (3106)

### Changed
- Changed pycrypto dependency with pycryptodome

## [1.0.2] - 2018-12-29
### Fixed
- Fixed code to support python3

## [1.0.1] - 2018-12-28

- First public release


midea_inventor_lib/midea_client.py
import time
import datetime
import requests
import json
import logging
from .midea_security import MideaSecurity
from .midea_databody_dehumi_query import DataBodyDeHumiQuery
from .midea_databody_dehumi_response import DataBodyDeHumiResponse
from .midea_dehumidification_device import MideaDehumidificationDevice
from .midea_databody_dehumi_request import DataBodyDeHumiRequest


#Main class
class MideaClient:

  APP_KEY = "3742e9e5842d4ad59c2db887e12449f9"
  APP_ID = 1017
  SRC = 17
  SERVER_URL  = "https://mapp.appsmb.com"
  CLIENT_TYPE = 1                 # Android
  FORMAT      = 2                 # JSON
  LANGUAGE    = "en_US"


  def __init__(self, email, password, sha256password, debug=False, verbose=False, logFile="", cacheTimeInSeconds=300):
    #Set logging
    if debug:
      loglevel=logging.DEBUG
      loglevelStr = "DEBUG"
    elif verbose:
      loglevel=logging.INFO
      loglevelStr = "INFO"
    else:
      loglevel=logging.WARNING
      loglevelStr = "WARNING"

    if not logFile:
      logging.basicConfig(level=loglevel)
      logging.info("MideaClient: logging to console.")
    else:
      logging.basicConfig(filename=logFile, level=loglevel)
      logging.info("MideaClient: logging to file: %s", logFile)

    logging.info("MideaClient: logging level set to: %s", loglevelStr)


    logging.debug("MideaClient: initializing MideaClient object")
    self.email    = email
    self.password = password
    self.sha256password = sha256password
    self.app_key  = self.APP_KEY
    self.app_id   = self.APP_ID
    self.src      = self.SRC

    logging.debug("MideaClient: creating MideaSecurity object using APPKEY=%s", self.app_key)
    self.security = MideaSecurity(self.app_key)

    self.current = None
    self.deviceStatus = None
    self.default_home = None
    self.cacheTimeStamp = 0    # Used to cache deviceStatus data
    self.CACHE_TIME = cacheTimeInSeconds


  def login(self):
    logging.debug("MideaClient: executing login()")

    #Receive loginId by calling '/v1/user/login/id/get' API endpoint
    response = self.__api_request("/v1/user/login/id/get", {"loginAccount": self.email})

    if "result" in response:
      result = response["result"]
      logging.debug("MideaClient::appliance_transparent_send: result=%s", json.dumps(result))
    else:
      logging.error("MideaClient::appliance_transparent_send: result=ERROR in API response")
      return -1

    if not "loginId" in result:
      logging.error("MideaClient: ERROR trying to get login ID.")
      return -1

    login_id = result["loginId"]
    logging.info("MideaClient: login ID successfully retrived: %s", login_id)

    if self.password:
      encrypted_password = self.security.loginEncrypt(self.password, login_id)
    elif self.sha256password:
      encrypted_password = self.security.loginEncryptWithSHA256password(self.sha256password, login_id)
    else:
      logging.error("MideaClient: ERROR: a plaintext password or sha256 hash must be provided to log in.")
      return -1

    #Log in to get valid accessToken and sessionId by calling '/v1/user/login' endpoint
    response = self.__api_request("/v1/user/login",
      {"loginAccount": self.email,
      "password": encrypted_password}
    )

    if "result" in response:
      self.current = response["result"]
      logging.debug("MideaClient::appliance_transparent_send: result=%s", json.dumps(result))
    else:
      logging.error("MideaClient::appliance_transparent_send: result=ERROR in API response")
      return -1

    if not "accessToken" in self.current:
      logging.error("MideaClient: ERROR trying to log in.")
      return -1

    self.security.access_token = self.current["accessToken"]
    logging.info("MideaClient: successfully logged in.")
    logging.info("MideaClient: accessToken=%s", self.security.access_token)
    logging.info("MideaClient: sessionId=%s", self.current["sessionId"])

    return 1


  def listAppliances(self):
    logging.debug("MideaClient: listAppliances()")

    if self.current is None:
      logging.warning("MideaClient::listAppliances: API session is not initialized: login first.")
      return None

    #Receive homegroupId by calling '/v1/homegroup/list/get' API endpoint
    response = self.__api_request("/v1/homegroup/list/get")

    if "result" in response:
      result = response["result"]
      logging.debug("MideaClient::appliance_transparent_send: result=%s", json.dumps(result))
    else:
      logging.error("MideaClient::appliance_transparent_send: result=ERROR in API response")
      return -1

    if not "list" in result:
      logging.error("MideaClient::listAppliances: ERROR getting devices list.")
      return None

    #result["list"] is an array of dictionaries
    self.default_home = None
    for dict in result["list"]:
      if dict["isDefault"] == "1":
        self.default_home = dict
        logging.info("MideaClient::listAppliances: homegroupId: %s", dict["id"])

    if self.default_home is None:
      logging.error("MideaClient::listAppliances: ERROR: homegroupId not found.")
      return None

    #Receive appliance info by calling '/v1/appliance/list/get' API endpoint
    response = self.__api_request("/v1/appliance/list/get",
        {"homegroupId": self.default_home["id"]}
      )

    if "result" in response:
      result = response["result"]
      logging.debug("MideaClient::appliance_transparent_send: result=%s", json.dumps(result))
    else:
      logging.error("MideaClient::appliance_transparent_send: result=ERROR in API response")
      return -1

    if not "list" in result:
      logging.error("MideaClient::listAppliances: ERROR getting devices list.")
      return None

    return result["list"]


  def get_device_status(self, deviceId):
    #Check if cached results are present and cab be used
    if self.deviceStatus is not None and self.cacheTimeStamp:
        now = time.time()
        if now - self.cacheTimeStamp < self.CACHE_TIME:
            #Cache results can be used
            logging.info("MideaClient::get_device_status (cached): %s", self.deviceStatus.toString())
            return 1


    if self.current is None:
      logging.warning("MideaClient::get_device_status: API session is not initialized: please login first.")
      return -1

    query = DataBodyDeHumiQuery()
    decoded_order = query.toBytes()

    #signed to unsigned bytes conversion
    decoded_order2 = []
    for i in decoded_order:
      if i < 0:
        decoded_order2.append(i+256)
      else:
        decoded_order2.append(i)

    #Header for switch-on/off & query-status command
    data = [90,90,1,0,89,0,32,0,1,0,0,0,39,36,17,9,13,10,18,20,218,73,0,0,0,16,0,0,0,0,0,0,0,0,0,0,0,0,0,0]

    data += decoded_order2
    data += [0]*16
    data[4] = len(data)

    #Convert array of int into a string (remove beginning and ending brackets)
    dataStr = ",".join(str(k) for k in data)
    decodedReplyStr = self.appliance_transparent_send(deviceId, dataStr)

    if decodedReplyStr is None:
      logging.error("MideaClient::get_device_status: invalid API response")
      return -1

    #Convert from comma-separated int string to int array
    decoded_reply = []
    for d in decodedReplyStr.split(","):
      n = int(d)
      decoded_reply.append(n)

    status = decoded_reply[40:]
    if not status:
      logging.error("MideaClient::get_device_status: got invalid status from API response")
      return -1

    #Process response (get device status)
    response = DataBodyDeHumiResponse()
    self.deviceStatus = response.toMideaDehumidificationDeviceObject(status)
    self.cacheTimeStamp = time.time()    #Update cacheTimeStamp

    logging.info("MideaClient::get_device_status: %s", self.deviceStatus.toString())
    return 1


  def send_poweron_command(self, deviceId):
    if self.current is None:
      logging.info("MideaClient::send_poweron_command: API session is not initialized: please login first.")
      return None

    if self.deviceStatus is None:
      logging.info("MideaClient::send_poweron_command: device's status unknown: please call get_device_status() first.")
      return None

    if self.deviceStatus.powerMode == 1:
      logging.info("MideaClient::send_poweron_command: device is already on.")
      return None

    #Create new command request
    request = DataBodyDeHumiRequest()
    request.setDataBodyStatus(self.deviceStatus)
    request.powerMode = 1  #power-on command

    #Header for switch-on/off & query-status command
    header = [90,90,1,0,89,0,32,0,1,0,0,0,39,36,17,9,13,10,18,20,218,73,0,0,0,16,0,0,0,0,0,0,0,0,0,0,0,0,0,0]
    return self.__send_command(request, header, deviceId)


  def send_poweroff_command(self, deviceId):
    if self.current is None:
      logging.info("MideaClient::send_poweroff_command: API session is not initialized: please login first.")
      return 0

    if self.deviceStatus is None:
      logging.info("MideaClient::send_poweroff_command: device's status unknown: please call get_device_status() first.")
      return 0

    if self.deviceStatus.powerMode == 0:
      logging.info("MideaClient::send_poweroff_command: device is already off.")
      return 0

    #Create new command request
    request = DataBodyDeHumiRequest()
    request.setDataBodyStatus(self.deviceStatus)
    request.powerMode = 0  #power-off command

    #Header for switch-on/off & query-status command
    header = [90,90,1,0,89,0,32,0,1,0,0,0,39,36,17,9,13,10,18,20,218,73,0,0,0,16,0,0,0,0,0,0,0,0,0,0,0,0,0,0]
    return self.__send_command(request, header, deviceId)


  def send_fan_speed_silent_command(self, deviceId):
    self.send_fan_speed_command(deviceId, 20)

  def send_fan_speed_medium_command(self, deviceId):
    self.send_fan_speed_command(deviceId, 60)

  def send_fan_speed_high_command(self, deviceId):
    self.send_fan_speed_command(deviceId, 80)


  def send_fan_speed_command(self, deviceId, speed):
    if self.current is None:
      loggin.info("MideaClient::send_fan_speed_command: API session is not initialized: please login first.")
      return 0

    if self.deviceStatus is None:
      logging.info("MideaClient::send_fan_speed_command: device's status unknown: please call get_device_status() first.")
      return 0

    if self.deviceStatus.powerMode == 0:
      logging.info("MideaClient::send_fan_speed_command: device is off.")
      return 0

    if not (speed > 0 and speed < 100):
      logging.info("MideaClient::send_fan_speed_command: speed value is not valid.")
      return 0

    #Create new command request
    request = DataBodyDeHumiRequest()
    request.setDataBodyStatus(self.deviceStatus)
    request.windSpeed = speed  #set speed

    #Header for non-power-related command
    header = [90,90,1,0,91,0,32,0,10,0,0,0,10,10,10,3,2,11,18,20,218,73,0,0,0,16,0,0,0,0,0,0,0,0,0,0,0,0,0,0]
    return self.__send_command(request, header, deviceId)


  def send_target_humidity_command(self, deviceId, humidity):
    if self.current is None:
      logging.info("MideaClient::send_target_humidity_command: API session is not initialized: please login first.")
      return None

    if self.deviceStatus is None:
      logging.info("MideaClient::send_target_humidity_command: device's status unknown: please call get_device_status() first.")
      return None

    if self.deviceStatus.powerMode == 0:
      logging.info("MideaClient::send_target_humidity_command: device is off.")
      return None

    if not (humidity >= 30 and humidity <= 70):
      logging.info("MideaClient::send_target_humidity_command: range for target humidity is not valid.")
      return None

    #Create new command request
    request = DataBodyDeHumiRequest()
    request.setDataBodyStatus(self.deviceStatus)
    request.humidity_set = humidity  #set target humidity

    #Header for non-power-related command
    header = [90,90,1,0,91,0,32,0,10,0,0,0,10,10,10,3,2,11,18,20,218,73,0,0,0,16,0,0,0,0,0,0,0,0,0,0,0,0,0,0]
    return self.__send_command(request, header, deviceId)


  def send_target_mode_command(self, deviceId):
    self.send_mode_command(deviceId, 1)

  def send_continous_mode_command(self, deviceId):
    self.send_mode_command(deviceId, 2)

  def send_smart_mode_command(self, deviceId):
    self.send_mode_command(deviceId, 3)

  def send_dryer_mode_command(self, deviceId):
    self.send_mode_command(deviceId, 4)


  def send_mode_command(self, deviceId, mode):
    if self.current is None:
      logging.info("MideaClient::send_mode_command: API session is not initialized: please login first.")
      return 0

    if self.deviceStatus is None:
      logging.info("MideaClient::send_mode_command: device's status unknown: please call get_device_status() first.")
      return 0

    if self.deviceStatus.powerMode == 0:
      logging.info("MideaClient::send_mode_command: device is off.")
      return 0

    if not (mode > 0 and mode < 5):
      logging.info("MideaClient::send_mode_command: mode is not valid.")
      return 0

    #Create new command request
    request = DataBodyDeHumiRequest()
    request.setDataBodyStatus(self.deviceStatus)
    request.setMode = mode  #set Mode (1:TARGET_MODE, 2:CONTINOUS_MODE, 3:SMART_MODE, 4:DRYER_MODE)

    #Header for non-power-related command
    header = [90,90,1,0,91,0,32,0,10,0,0,0,10,10,10,3,2,11,18,20,218,73,0,0,0,16,0,0,0,0,0,0,0,0,0,0,0,0,0,0]

    return self.__send_command(request, header, deviceId)


  def send_ion_on_command(self, deviceId):
    if self.current is None:
      logging.info("MideaClient::send_ion_on_command: API session is not initialized: please login first.")
      return 0

    if self.deviceStatus is None:
      logging.info("MideaClient::send_ion_on_command: device's status unknown: please call get_device_status() first.")
      return 0

    if self.deviceStatus.powerMode == 0:
      logging.info("MideaClient::send_ion_on_command: device is off.")
      return 0

    if self.deviceStatus.ionSetSwitch == 1:
      logging.info("MideaClient::send_ion_on_command: Ion mode is alreay on.")
      return 0

    #Create new command request
    request = DataBodyDeHumiRequest()
    request.setDataBodyStatus(self.deviceStatus)
    request.ionSetSwitch = 1  #on

    #Header for non-power-related command
    header = [90,90,1,0,91,0,32,0,10,0,0,0,10,10,10,3,2,11,18,20,218,73,0,0,0,16,0,0,0,0,0,0,0,0,0,0,0,0,0,0]

    return self.__send_command(request, header, deviceId)


  def send_ion_off_command(self, deviceId):
    if self.current is None:
      logging.info("MideaClient::send_ion_off_command: API session is not initialized: please login first.")
      return 0

    if self.deviceStatus is None:
      logging.info("MideaClient::send_ion_off_command: device's status unknown: please call get_device_status() first.")
      return 0

    if self.deviceStatus.powerMode == 0:
      logging.info("MideaClient::send_ion_off_command: device is off.")
      return 0

    if self.deviceStatus.ionSetSwitch == 0:
      logging.info("MideaClient::send_ion_off_command: Ion mode is alreay off.")
      return 0

    #Create new command request
    request = DataBodyDeHumiRequest()
    request.setDataBodyStatus(self.deviceStatus)
    request.ionSetSwitch = 0  #off

    #Header for non-power-related command
    header = [90,90,1,0,91,0,32,0,10,0,0,0,10,10,10,3,2,11,18,20,218,73,0,0,0,16,0,0,0,0,0,0,0,0,0,0,0,0,0,0]

    return self.__send_command(request, header, deviceId)


  def send_update_status_command(self, deviceId, deviceStatus):
    """ Generic command to set a new device status """

    if self.current is None:
      logging.info("MideaClient::send_ion_off_command: API session is not initialized: please login first.")
      return 0

    if self.deviceStatus.powerMode == 0:
      logging.info("MideaClient::send_ion_off_command: device is off.")
      return 0

    #Sanity checks
    if deviceStatus.powerMode != 0 and deviceStatus.powerMode != 1:
      logging.error("MideaClient::send_udpate_statis_command: invalid powerMode specified in device status.")
      return 0
    if not (deviceStatus.setMode >0 and deviceStatus.setMode <5):
      logging.error("MideaClient::send_udpate_statis_command: invalid setMode specified in device status.")
      return 0
    if not (deviceStatus.humidity_set >=30 and deviceStatus.humidity_set <=70):
      logging.error("MideaClient::send_udpate_statis_command: invalid target humidity specified in device status.")
      return 0
    #if not (deviceStatus.humidity_dot_set >=30 and deviceStatus.humidity_dot_set <=70):
    #  logging.error("MideaClient::send_udpate_statis_command: invalid target humidity decimal specified in device status.")
    #  return 0
    if not (deviceStatus.windSpeed > 0 and deviceStatus.windSpeed <100):
      logging.error("MideaClient::send_udpate_statis_command: invalid windSpeed specified in device status.")
      return 0
    if deviceStatus.ionSetSwitch != 0 and deviceStatus.ionSetSwitch != 1:
      logging.error("MideaClient::send_udpate_statis_command: invalid ionSetSwitch specified in device status.")
      return 0


    #Create new command request
    request = DataBodyDeHumiRequest()
    request.setDataBodyStatus(deviceStatus)

    #Header for non-power-related command
    header = [90,90,1,0,91,0,32,0,10,0,0,0,10,10,10,3,2,11,18,20,218,73,0,0,0,16,0,0,0,0,0,0,0,0,0,0,0,0,0,0]

    return self.__send_command(request, header, deviceId)


  def __send_command(self, request, header, deviceId):
    decoded_order = request.toBytes()

    #signed to unsigned bytes conversion
    decoded_order2 = []
    for i in decoded_order:
      if i < 0:
        decoded_order2.append(i+256)
      else:
        decoded_order2.append(i)

    data = header
    data += decoded_order2
    data += [0]*16
    data[4] = len(data)

    #Convert array of int into a string (remove beginning and ending brackets)
    dataStr = ",".join(str(k) for k in data)
    decodedReplyStr = self.appliance_transparent_send(deviceId, dataStr)

    #Convert from comma-separated int string to int array
    decoded_reply = []
    for d in decodedReplyStr.split(","):
      n = int(d)
      decoded_reply.append(n)

    status = decoded_reply[40:]
    #Process response (get device status)
    response = DataBodyDeHumiResponse()
    self.deviceStatus = response.toMideaDehumidificationDeviceObject(status)
    logging.info("MideaClient::__send_command: %s", self.deviceStatus.toString())
    return 1



  def appliance_transparent_send(self, appliance_id, dataStr):
    """dataStr is a string of comma-separated integer"""
    dataArr = self.security.transcode(dataStr)
    dataStr2 = ",".join(str(k) for k in dataArr)

    encoded_command = self.__encode(dataStr2)
    logging.debug("MideaClient::appliance_transparent_send: encoded_command=%s", encoded_command)

    #Send command to applicance via API /v1/appliance/transparent/send endpoint
    response = self.__api_request("/v1/appliance/transparent/send",
        { "order": encoded_command,
          "funId": "0000",
          "applianceId": appliance_id })


    errorCodeStr = ""
    if "errorCode" in response:
      errorCodeStr = response["errorCode"]

    if errorCodeStr != "0":
      logging.error("MideaClient::send_api_request: errorCode=%s, errorMessage=\"%s\"", response["errorCode"], response["msg"])
      if errorCodeStr == "3106":
        #Manage invalidSession Error by performing a new login
        logging.error("MideaClient: trying to login again due to invalidSession error...")
        self.login()

        #Re-encode command after login
        encoded_command = self.__encode(dataStr2)
        logging.debug("MideaClient::appliance_transparent_send: encoded_command=%s", encoded_command)

        #Send again the command to applicance via API /v1/appliance/transparent/send endpoint
        response = self.__api_request("/v1/appliance/transparent/send",
            { "order": encoded_command,
              "funId": "0000",
              "applianceId": appliance_id })


    if "result" in response:
      result = response["result"]
      logging.debug("MideaClient::appliance_transparent_send: result=%s", json.dumps(result))
      if not "reply" in result:
        logging.error("MideaClient::appliance_transparent_send: result=ERROR in API response: reply key not found")
        return None
      else:
        return self.__decode(result["reply"])
    else:
      logging.error("MideaClient::appliance_transparent_send: result=ERROR in API response: result key not found")
      return None


  def __api_request(self, endpoint, _args={}):
    ts = time.time()
    st = datetime.datetime.fromtimestamp(ts).strftime('%Y%m%d%H%M%S')

    args = {}
    args["appId"] = self.app_id
    args["format"] = self.FORMAT
    args["clientType"] = self.CLIENT_TYPE
    args["language"] = self.LANGUAGE
    args["src"] = self.src
    args["stamp"] = int(st)
    args.update(_args)

    if self.current is not None:
      args["sessionId"] = str(self.current["sessionId"])

    argsStr = "&".join("=".join((str(k),str(v))) for k,v in args.items())
    #print argsStr
    args["sign"] = self.security.sign(endpoint, argsStr)

    #Call __send_api_request
    path = self.SERVER_URL+endpoint
    response = self.__send_api_request(path, args)

    return response


  def __send_api_request(self, uri, args):
    #Https POST request
    response = requests.post(uri, data=args)
    logging.debug("MideaClient::send_api_request: response=%s", response.text) #TEXT/HTML
    logging.info("MideaClient::send_api_request: response_status=%s, response_reason=%s", response.status_code, response.reason) #HTTP

    data = response.json()
    return data


  def __encode(self, data):
    return self.security.aes_encrypt(data, self.security.data_key)


  def __decode(self, data):
    return self.security.aes_decrypt(data, self.security.data_key)
